ruby_programming/git/advanced_git_lesson.md

```diff
@@ -0,0 +1,182 @@
+5) Regarding the specific commands we've covered:
+  - For `git amend` never amend commits that have been pushed to remote repositories
+  - For `git rebase` never rebase a repository that others may work off of 
+  - For `git reset` never reset commits thath have been pushed to remote repositories
+  - For `git push --force` only use it when appropriate, use it with caution, and      preferably default to using `git push --force-with-lease`
+
+### Branches Are Pointers
+
+While the focus of this lesson was more advanced tools for changing Git history, we're going into another advanced topic that might be hard for some to understand - Pointers. You've already learned about [branches](https://www.theodinproject.com/paths/foundations/courses/foundations/lessons/revisiting-rock-paper-scissors) and how these hold multiple *alternate reality* versions of our files. Now we're going to discuss what that actually means under the hood, and what it means for branches to be pointers.
+
+Before we dive into branches, let's talk about commits. If we recall [first learning about commits](https://www.theodinproject.com/paths/foundations/courses/foundations/lessons/git-basics), they were described as Snapshots. If it helps, think of this in a very literal sense. Every time you type in `git commit` your computer is taking a picture of all the file contents that have been staged with `git add`, in other words your entire tracked workspace gets copied. 
+
+So what is a branch? Based off of your exposure, you might be visualizing a branch as a group of commits. This actually isn't the case! **A branch is actually a pointer to a single commit!** Hearing this your first thought might be *"Well if a branch is just a finger pointing at a single commit, how does that single commit know about all the commits that came before it?"*. The answer to the question is very simple, each commit is also a pointer, and points to the commit that came before it! Wow. This might be a lot to take in, so let's take a moment to absorb that fact. 
+
+Now that you've had a second to gather your thoughts and attempt to wrap your head around this concept, it might help to go back and look at a concrete example of pointers we used in this lesson. Let's think back to our use of `git rebase -i HEAD~3`. If you can remember, this command let us edit the last 3 commits. Do you have any guesses on how git knew which 3 commits to edit? That's right, by using pointers! We start at HEAD, a special pointer for keeping track of the branch you're currently on. HEAD points to our most recent commit in the current branch. That commit pointed to the commit made directly before it, which we can call commit 2. That commit does the exact same, and points to the commit before it, which we can call commit 3. That's how `git rebase -i HEAD~3` starts with a HEAD pointer, and then uses more pointers to see which three commits to edit. 
```

Comment:
```suggestion
Now that you've had a second to gather your thoughts and attempt to wrap your head around this concept, it might help to go back and look at a concrete example of pointers we used in this lesson. Let's think back to our use of `git rebase -i HEAD~3`. If you can remember, this command lets us edit the last 3 commits. Do you have any guesses on how git knew which 3 commits to edit? That's right, by using pointers! We start at HEAD, a special pointer for keeping track of the branch you're currently on. HEAD points to our most recent commit in the current branch. That commit points to the commit made directly before it, which we can call commit 2. That commit does the exact same, and points to the commit before it, which we can call commit 3. That's how `git rebase -i HEAD~3` starts with a HEAD pointer, and then uses more pointers to see which three commits to edit. 
```